pkg/graph/dag/dag.go

```diff
@@ -42,134 +56,128 @@ func NewDirectedAcyclicGraph() *DirectedAcyclicGraph {
+		ID:        id,
+		Order:     order,
+		DependsOn: make(map[string]struct{}),
 	}
 	return nil
 }
 
 type CycleError struct {
-	From, to string
-	Cycle    []string
+	Cycle []string
 }
 
 func (e *CycleError) Error() string {
-	return fmt.Sprintf("Cannot add edge from %s to %s. This would create a cycle: %s", e.From, e.to, formatCycle(e.Cycle))
+	return fmt.Sprintf("graph contains a cycle: %s", formatCycle(e.Cycle))
 }
 
 func formatCycle(cycle []string) string {
 	return strings.Join(cycle, " -> ")
 }
 
+// AsCycleError returns the (potentially wrapped) CycleError, or nil if it is not a CycleError.
+func AsCycleError(err error) *CycleError {
+	cycleError := &CycleError{}
+	if errors.As(err, &cycleError) {
+		return cycleError
+	}
+	return nil
+}
+
 // AddEdge adds a directed edge from one node to another.
-func (d *DirectedAcyclicGraph) AddEdge(from, to string) error {
+// The edge indicates that "to" is a dependency of "from", i.e. "to" must occur before "from"
+func (d *DirectedAcyclicGraph) AddDependencies(from string, dependencies []string) error {
 	fromNode, fromExists := d.Vertices[from]
-	_, toExists := d.Vertices[to]
 	if !fromExists {
 		return fmt.Errorf("node %s does not exist", from)
 	}
-	if !toExists {
-		return fmt.Errorf("node %s does not exist", to)
-	}
-	if from == to {
-		return fmt.Errorf("self references are not allowed")
-	}
 
-	fromNode.Edges[to] = struct{}{}
+	for _, dependency := range dependencies {
+		_, toExists := d.Vertices[dependency]
+		if !toExists {
+			return fmt.Errorf("node %s does not exist", dependency)
+		}
+		if from == dependency {
+			return fmt.Errorf("self references are not allowed")
+		}
+		fromNode.DependsOn[dependency] = struct{}{}
+	}
 
 	// Check if the graph is still a DAG
-	hasCycle, cycle := d.HasCycle()
+	hasCycle, cycle := d.hasCycle()
 	if hasCycle {
 		// Ehmmm, we have a cycle, let's remove the edge we just added
-		delete(fromNode.Edges, to)
+		for _, dependency := range dependencies {
+			delete(fromNode.DependsOn, dependency)
+		}
 		return &CycleError{
-			From:  from,
-			to:    to,
 			Cycle: cycle,
 		}
 	}
 
 	return nil
 }
 
+// TopologicalSort returns the vertexes of the graph, respecting topological ordering first,
+// and preserving order of nodes within each "depth" of the topological ordering.
 func (d *DirectedAcyclicGraph) TopologicalSort() ([]string, error) {
-	if cyclic, _ := d.HasCycle(); cyclic {
-		return nil, fmt.Errorf("graph has a cycle")
-	}
-
 	visited := make(map[string]bool)
 	var order []string
 
-	// Get a sorted list of all vertices
-	vertices := d.GetVertices()
+	// Make a list of vertices, sorted by Order
+	vertices := make([]*Vertex, 0, len(d.Vertices))
+	for _, vertex := range d.Vertices {
+		vertices = append(vertices, vertex)
+	}
+	sort.Slice(vertices, func(i, j int) bool {
+		return vertices[i].Order < vertices[j].Order
+	})
 
-	var dfs func(string)
-	dfs = func(node string) {
-		visited[node] = true
+	for len(visited) < len(vertices) {
```

Comment:
This algorithm is less efficient, but more straightforward and I think can detect cycles also (so we might be able to remove cycle detection from adding the edges).